src/enochecker/nosqldict.py

```diff
@@ -267,14 +276,18 @@ def persist(self) -> None:
         """
         Persist the changes in the backend.
         """
-        for (key, value) in self.cache.items():
-            hash_ = value_to_hash(value)
-            if (
-                (not hash_)
-                or (key not in self.hash_cache)
-                or (self.hash_cache[key] != hash_)
-            ):
-                self._upsert(key, value)
+        with self._lock:
+            for (key, value) in list(self.cache.items()):
+                hash_ = value_to_hash(value)
+                if (
+                    (not hash_)
+                    or (key not in self.hash_cache)
+                    or (self.hash_cache[key] != hash_)
+                ):
+                    self._upsert(key, value)
+                del self.cache[key]
```

Comment:
Won't this change the dict?

Reply:
Yes, but calling `list(self.cache.items())` empties the `items()`-iterator *before* the first iteration of the loop. The original issue was caused by chaning the dict whlie still consuming from the `items()`-iterator